---
title: "DESeq2 Course Work"
author: "Dr. Shirin Glander"
date: "`r Sys.Date()`"
output:
  md_document:
    variant: markdown_github
---

# DESeq2 Course Work

Go to [exprAnalysis](https://github.com/ShirinG/exprAnalysis/) for installation instructions.

For all functions, use the help pages to find out more about parameters and usage.

See Vignette for additional information.

```{r library}
library(exprAnalysis)
```

-----------

<br>

## Input data

> "As input, the DESeq2 package expects count data as obtained, e. g., from RNAseq
or another high-throughput sequencing experiment, in the form of a matrix of
integer values. The value in the i-th row and the j-th column of the matrix tells
how many reads can be assigned to gene i in sample j." Love et al., DESeq2 vignette

```{r countmatrix}
data("countmatrix")
```

-----------

<br>

## Count data analysis with DESeq2

See [DESeq2 Vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf) for details.

- read in saved count matrix
- define experimental design
- convert to DESeq data set

#### Count matrix input

```{r DESeqDataSetFromMatrix, cache=TRUE}
design <- gsub("(.*)(_[0-9])", "\\1", colnames(countmatrix))
ExpDesign <- data.frame(row.names=colnames(countmatrix), treatment = design)

data <- DESeq2::DESeqDataSetFromMatrix(countData = countmatrix, colData = ExpDesign, design = ~treatment)
```

-----------

<br>

### DESeq2

- optional, but recommended: remove genes with zero counts over all samples
- run DESeq
- Extracting transformed values

> "While it is not necessary to pre-filter low count genes before running the DESeq2
functions, there are two reasons which make pre-filtering useful: by removing rows
in which there are no reads or nearly no reads, we reduce the memory size of
the dds data object and we increase the speed of the transformation and testing
functions within DESeq2." Love et al., DESeq2 vignette

Note: the rlog transformation is provided for applications other than differential testing. For differential testing we recommend the DESeq function applied to raw counts, as described later in this workflow, which also takes into account the dependence of the variance of counts on the mean value during the dispersion estimation step.

For a quick first glance at the data, we can use <span style="color:red">pcaExplorer</span>.

```{r DESeq, cache=TRUE}
data <- data[rowSums(DESeq2::counts(data)) > 1, ]

data_DESeq <- DESeq2::DESeq(data)

expmatrix_DESeq <- DESeq2::rlog(data_DESeq, fitType="local")
expmatrix <- SummarizedExperiment::assay(expmatrix_DESeq)
```

```{r pcaExplorer, eval=FALSE}
library("pcaExplorer")
pcaExplorer(data_DESeq, expmatrix_DESeq)
```

-----------

<br>

### Dispersion plot

```{r dispersion_plot, fig.width=6, fig.height=5}
DESeq2::plotDispEsts(data_DESeq, main="Dispersion Estimates")
```

-----------

<br>

## Exploratory analysis of all genes

### Variance vs mean gene expression across samples

Plots variance against mean gene expression across samples and calculates the correlation of a linear regression model.

**var_vs_mean()** uses the R package matrixStats.

```{r var_vs_mean, fig.width=6, fig.height=5, cache=TRUE}
var_vs_mean(countmatrix)
var_vs_mean(expmatrix)
```

-----------

<br>

### Intersample variances

```{r variance_sumOverlaps, fig.width= 5, cache=TRUE}
library(corrgram)

Ctrl_cor <- expmatrix[,grep("Ctrl", colnames(expmatrix))]

corrgram::corrgram(Ctrl_cor, order=TRUE, lower.panel=corrgram::panel.pie,
         upper.panel=corrgram::panel.pts, text.panel=corrgram::panel.txt,
         main="Correlogram of controls")
```

#### Repeat for other treatment groups

-----------

<br>

### Principle Component Analysis

Uses functions from the R package pcaGoPromoter.

You can only plot the principle components using:

```{r pca_plot, fig.width=6, fig.height=6, cache=TRUE}
groups <- as.factor(c(rep("Ctrl",4), rep("TolLPS",4), rep("TolS100A8",4), rep("ActLPS",4)))
pca_plot(expmatrix, groups)
```

<br>

Or you can plot the principle components and calculate TF and GO term enrichments of genes (defaults to top 2.5%) with highest and lowest loadings. With this function, the ouput files are directly saved to .pdf and .txt (by default to working directory).

```{r pca_plot_enrich, eval=FALSE}
pca_plot_enrich(expmatrix, groups)
```

-----------

<br>

### Heatmaps

**heatmaps()** uses the R package gplots.

Here, of the 30 most highly expressed genes.

```{r heatmaps, cache=TRUE}
select <- order(rowMeans(expmatrix),decreasing=TRUE)[1:30]
heatmaps(expmatrix[select,], samplecols = rep(c("#E41A1C", "#377EB8", "#4DAF4A", "#984EA3"), each=4))
```

#### Heatmap function from DESeq2, using pheatmap:

```{r heatmaps2, cache=TRUE}
library(pheatmap)

sampleDists <- dist(t(expmatrix))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(expmatrix_DESeq$treatment)
colnames(sampleDistMatrix) <- NULL
colors <- grDevices::colorRampPalette( rev(RColorBrewer::brewer.pal(9, "Blues")) )(255)
pheatmap::pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

df <- data.frame(treatment = SummarizedExperiment::colData(data_DESeq)[,c("treatment")], row.names = rownames(SummarizedExperiment::colData(data_DESeq)))
pheatmap::pheatmap(expmatrix[select,], cluster_rows=TRUE, show_rownames=TRUE, cluster_cols=TRUE, annotation_col=df)
```

-----------

<br>

### Hierarchical Clustering and outlier detection

Uses adjacency matrix function from the R package WGCNA and hierarchical clustering from the R package flashClust.

```{r wgcna_sample_dendrogram, fig.width=8, fig.height=5, cache=TRUE}
datTraits <- data.frame(Ctrl = c(rep(1, 4), rep(0,12)), TolPS = c(rep(0, 4), rep(1, 4),rep(0, 8)), TolS100A8 = c(rep(0, 8), rep(1, 4), rep(0, 4)), ActLPS = c(rep(0, 12),rep(1, 4)), Tol = c(rep(0, 4), rep(1, 8), rep(0, 4)), ExPhenotype = c(stats::rnorm(4, 10, 1),stats::rnorm(8, 25, 1),stats::rnorm(4, 50, 1)), row.names = colnames(expmatrix))

datExpr <- wgcna_sample_dendrogram(expmatrix, datTraits)

# Optional: Remove outlier samples and repeats: All genes flagged for removal are saved to the object "remove_genes"
#head(remove_genes)
```

-----------

<br>

## Differential expression analysis using DESeq2

For raw read count data.

##### contrast DE groups:

 + lfc = treatment > Ctrl, - lfc = treatment < Ctrl
 p-value & p.adjust values of NA indicate outliers detected by Cook's distance
 NA only for p.adjust means the gene is filtered by automatic independent filtering for having a low mean normalized count

Information about which variables and tests were used can be found by calling the
function  <span style="color:red">mcols</span>,  on the results object.

```{r DEgenes_DESeq2, cache=TRUE}
library(DESeq2)
library(ggplot2)
library(ggrepel)

# find possible contrasts with
DESeq2::resultsNames(data_DESeq)

res <- DESeq2::results(data_DESeq, contrast=list("treatmentActLPS", "treatmentCtrl"), cooksCutoff = 0.99, independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH")
summary(res)
mcols(res)$description

# order results table by the smallest adjusted p value:
res <- res[order(res$padj),]

results = as.data.frame(dplyr::mutate(as.data.frame(res), sig=ifelse(res$padj<0.05, "FDR<0.05", "Not Sig")), row.names=rownames(res))
head(results)

DEgenes_DESeq <- results[which(abs(results$log2FoldChange) > log2(1.5) & results$padj < 0.05),]

p = ggplot2::ggplot(results, ggplot2::aes(log2FoldChange, -log10(pvalue))) +
  ggplot2::geom_point(ggplot2::aes(col = sig)) +
  ggplot2::scale_color_manual(values = c("red", "black")) +
  ggplot2::ggtitle("Volcano Plot of DESeq2 analysis")

p + ggrepel::geom_text_repel(data=results[1:10, ], ggplot2::aes(label=rownames(results[1:10, ])))

# If there aren't too many DE genes:
#p + geom_text_repel(data = dplyr::filter(results, padj<0.05), aes(label = rownames(results[1:10, ])))

```

-----------

<br>

#### MA-plot

> "These plots show the log2 fold changes from the treatment over the mean
of normalized counts, i.e. the average of counts normalized by size factors. The left plot shows
the "unshrunken" log2 fold changes, while the right plot, produced by the code above, shows
the shrinkage of log2 fold changes resulting from the incorporation of zero-centered normal
prior. The shrinkage is greater for the log2 fold change estimates from genes with low counts
and high dispersion, as can be seen by the narrowing of spread of leftmost points in the right
plot." Love et al., DESeq2 vignette

```{r ma_plot, fig.width=5, fig.height=5, cache=TRUE}
DESeq2::plotMA(res, main="MA Plot", ylim=c(-2,2))
```

<br>

#### plotCounts

> "It can also be useful to examine the counts of reads for a single gene across the
groups. A simple function for making this plot is plotCounts, which normalizes
counts by sequencing depth and adds a pseudocount of 1/2
to allow for log scale
plotting. The counts are grouped by the variables in intgroup, where more than
one variable can be specified." Love et al., DESeq2 vignette

```{r plot_counts, fig.width=10, fig.height=3, cache=TRUE}
par(mfrow=c(1,3))

for (i in 1:3){
  gene <- rownames(res)[i]
  main = gene
  DESeq2::plotCounts(data_DESeq, gene=gene, intgroup="treatment", main = main)
}
```

-----------

<br>

## Gene annotations

Can be used to add e.g. ENTREZ ID, ENSEMBL ID, etc. to gene name.

```{r geneAnnotations, cache=TRUE}
results_anno <- geneAnnotations(input=results, keys=row.names(results), column=c("ENTREZID", "ENSEMBL"), keytype="SYMBOL", organism = "human")
head(results_anno)
```

-----------

<br>

## Enrichment Analysis using clusterPofiler

See [clusterProfiler](https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html) instructions for details.

```{r clusterProfiler, cache=TRUE}
library(clusterProfiler)
library(org.Hs.eg.db)
OrgDb <- org.Hs.eg.db # can also be other organisms

geneList <- as.vector(results_anno$log2FoldChange)
names(geneList) <- results_anno$ENTREZID
gene <- na.omit(results_anno$ENTREZID)


# Group GO
ggo <- clusterProfiler::groupGO(gene     = gene,
                                OrgDb    = OrgDb,
                                ont      = "BP",
                                level    = 3,
                                readable = TRUE)
head(summary(ggo)[,-5])
barplot(ggo, drop=TRUE, showCategory=12)
```

```{r clusterProfilerGO, fig.width = 10, fig.height = 5, cache=TRUE}
# GO over-representation test
ego <- clusterProfiler::enrichGO(gene          = gene,
                                 OrgDb         = OrgDb,
                                 ont           = "BP",
                                 pAdjustMethod = "BH",
                                 pvalueCutoff  = 0.05,
                                 qvalueCutoff  = 0.05, 
                                 readable      = TRUE)
head(summary(ego)[,-8])
barplot(ego, showCategory=25)
clusterProfiler::dotplot(ego, showCategory=25)
#clusterProfiler::plotGOgraph(ego)
```

```{r clusterProfilerKEGG, fig.width = 20, fig.height = 20, cache=TRUE}
## KEGG over-representation test
kk <- clusterProfiler::enrichKEGG(gene         = gene,
                 organism     = 'hsa',
                 pAdjustMethod = "BH",
                 pvalueCutoff = 0.05,
                 qvalueCutoff  = 0.05)
head(summary(kk)[,-8])
cnetplot(kk, categorySize="geneNum", foldChange=geneList)
```

