---
title: "Exploring the human genome (Part 3) - Chromosomes"
author: "Dr. Shirin Glander"
date: '`r Sys.Date()`'
output:
  prettydoc::html_pretty:
    theme: tactile
    highlight: github
---

In this part I want to explore the chromosomes of the human genome. We have 23 chromose pairs, 22 autosomal pairs which contain essentially the same genes (albeit in slightly different versions, called alleles) and the gonosomes (or sex chromosomes) X and Y, plus the mitochondrial (MT) chromosome. X and Y contain different genes and in this capacity encode sex-related traits. The MT is the remaining genetic information from what initially started as a symbiosis. The majority of the mitochondrial genes were lost over time or have been incorporated into the main nuclear DNA, but a handful of genes remains to this day. MT genes are interesting because they are solely inherited via the female line as they come with the mitochondria in the fertilized egg.

I am again using the **AnnotationDbi** and **EnsDb.Hsapiens.v79** packages. This time I want to extract the following information for each gene:

- on which chromosome it lives
- the length of the chromosome it lives on (in base pairs, bp)
- whether it is on the plus or minus strand
- and the coordinates of start and end position.

Based on start and end position of each gene I can calculate its length in base pairs (bp).

- *link to part 1 and 2*
- *link to EnsDb.Hsapiens.v79*
- *link to AnnotationDbi*

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8, fig.align="center", cache=TRUE}
library(AnnotationDbi)
library(EnsDb.Hsapiens.v79)
edb <- EnsDb.Hsapiens.v79

# get all gene IDs (Ensembl gene ID)
keys <- keys(edb, keytype = "GENEID")

chromosomes <- AnnotationDbi::select(edb, keys = keys, columns = c("SEQLENGTH", "SEQNAME", "SEQSTRAND", "GENESEQSTART", "GENESEQEND"), keytype = "GENEID")

# convert chromosome and strand to factors
chromosomes$SEQNAME <- as.factor(chromosomes$SEQNAME)
chromosomes$SEQSTRAND <- as.factor(chromosomes$SEQSTRAND)

# calculate gene length
chromosomes$GENELENGTH <- chromosomes$GENESEQEND - chromosomes$GENESEQSTART

head(chromosomes)
```

<br>

## How many bases of each chromosome are taken up by genes?

In order to calculate the sum of gene lengths for each chromosome I am using the *aggregate()* function. We can already see that there are many more unique chromosome identifiers than the 25 regular chromosomes we had expected:

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8, fig.align="center", cache=TRUE}
gene_bases <- aggregate(chromosomes$GENELENGTH, by = list(Chromosome = chromosomes$SEQNAME), FUN = sum)
length(unique(gene_bases$Chromosome))
```

This is because some of the genes have not yet been successfully mapped to a chromosome. Genes with seqnames that start with "KI" or "GL" for example, currently live on scaffolds where the chromosomal location is not yet known.

- **check what the other seqnames mean**

Here I only want to have a look at the 25 chromosomes and lump the unmapped scaffolds into one group of unknown chromosomes.

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8, fig.align="center", cache=TRUE}
chrom <- as.character(gene_bases[grep("^[0-9]+$|^X$|^Y$|^MT$", gene_bases$Chromosome), "Chromosome"])

gene_bases_other <- gene_bases[-which(gene_bases$Chromosome %in% chrom), ]
scaffolds <- data.frame(Chromosome = "S", x = sum(gene_bases_other[grep("^KI[0-9]+|^GL[0-9]+", gene_bases_other$Chromosome), "x"]))

library(dplyr)
gene_bases <- gene_bases[which(gene_bases$Chromosome %in% chrom), ] %>%
  rbind(scaffolds)

length(unique(gene_bases$Chromosome))
```

<br>

## How many genes are on each chromosome?

To count the number of genes per chromosome I am using the *table()* function.

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8, fig.align="center", cache=TRUE}
chromosomes_num_table <- as.data.frame(table(chromosomes$SEQNAME))

chromosomes_num_table_other <- chromosomes_num_table[-which(chromosomes_num_table$Var1 %in% chrom), ]
scaffolds <- data.frame(Var1 = "S", Freq = sum(chromosomes_num_table_other[grep("^KI[0-9]+|^GL[0-9]+", chromosomes_num_table_other$Var1), "Freq"]))

chromosomes_num_table <- chromosomes_num_table[which(chromosomes_num_table$Var1 %in% chrom), ] %>%
  rbind(scaffolds)
```

<br>

## How long are the chromosomes?

The information about chromosome length in base pairs (bp) is given in the column "SEQLENGTH":

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8, fig.align="center", cache=TRUE}
chrom_length <- chromosomes[!duplicated(chromosomes$SEQNAME), c(3,2)] #SEQNAME & SEQLENGTH

chrom_length_other <- chrom_length[-which(chrom_length$SEQNAME %in% chrom), ]
scaffolds <- data.frame(SEQNAME = "S", SEQLENGTH = sum(chrom_length_other[grep("^KI[0-9]+|^GL[0-9]+", chrom_length_other$SEQNAME), "SEQLENGTH"]))

chrom_length <- chrom_length[which(chrom_length$SEQNAME %in% chrom), ] %>%
  rbind(scaffolds)
```

<br>

Now, I can combine the information about

- how long each chromosome is,
- how much of it is made up of genes and
- how many genes there are it

into a single dataframe.

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8, fig.align="center", cache=TRUE}
chromosomes_table <- left_join(chrom_length, chromosomes_num_table, by = c("SEQNAME" = "Var1")) %>%
  left_join(gene_bases, by = c("SEQNAME" = "Chromosome"))

colnames(chromosomes_table) <- c("Chromosome", "Chr_length", "No_genes", "Gene_lengths_sum")
```

Based on this information I am calculating

- the number of genes per chromosome corrected for gene length and
- the percentage of each chromosome that is taken up by gene sequences.

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8, fig.align="center", cache=TRUE}
chromosomes_table$No_genes_by_Chr_length <- chromosomes_table$No_genes/chromosomes_table$Chr_length
chromosomes_table$Percent_of_Chr_covered_by_genes <- round(chromosomes_table$Gene_lengths_sum/chromosomes_table$Chr_length*100, digits = 2)
```

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8, fig.align="center", cache=TRUE}
library(reshape2)
chromosomes_table_m <- melt(chromosomes_table[, c(1,2,3,6)])

f = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y", "MT", "S")
chromosomes_table_m <- within(chromosomes_table_m, Chromosome <- factor(Chromosome, levels = f))

labels <- c(No_genes = "Number of Genes", 
            Chr_length = "Chromosome Length (in bp)", 
            Percent_of_Chr_covered_by_genes = "Percent of Chromosome\nCovered by Genes")

library(ggplot2)
p <- ggplot(data = chromosomes_table_m, aes(x = Chromosome, y = value)) + 
  geom_bar(stat = "identity", fill = "maroon", alpha = 0.8) + 
  theme(
    axis.text = element_text(size = 14),
    axis.title = element_text(size = 14),
    panel.grid.major = element_line(colour = "grey"),
    panel.grid.minor = element_blank(),
    panel.background = element_rect(fill = "white"),
    strip.background = element_rect(fill = "cornsilk", color = "maroon", size = 2),
    strip.text = element_text(face = "bold", size = 10, colour = "maroon")
  ) +
  labs(
    title = "Gene-Coverage of Chromosomes",
    x = "Chromosome", 
    y = "Value") +
  facet_grid(variable ~ ., scales = "free", labeller = labeller(variable = labels))

ann_text <- data.frame(x = chromosomes_table_m$Chromosome,
                       y = c(rep(55000000, 26), rep(600, 26), rep(26, 26)),
                       lab = as.character(chromosomes_table_m$value),
                       variable = chromosomes_table_m$variable)

p + geom_text(data = ann_text, aes(x, y, label = lab, group = variable), size = 4, angle = 90)
```

------------------

<br>

```{r echo=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=4, fig.align="center", cache=FALSE}
sessionInfo()
```

